### PR TITLE
HV-1277 Improve the performances of HV 6

### PR DESCRIPTION
**Pull request by gsmet**
* https://hibernate.atlassian.net/browse/HV-1277

This is a base of discussion. I already know you won't like some of the changes but it will at least allow to discuss them.

The second commit is what makes a BIG difference.

### CODE REVIEW

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java [hunk 1]
```diff
@@ -369,7 +372,7 @@ private ValidationOrder determineGroupValidationOrder(Class<?>[] groups) {
 			return Collections.emptySet();
 		}
 
-		BeanMetaData<U> beanMetaData = beanMetaDataManager.getBeanMetaData( valueContext.getCurrentBeanType() );
+		BeanMetaData<U> beanMetaData = valueContext.getCurrentBeanMetaData();
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java]: Nice.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java [hunk 2]
```diff
@@ -93,6 +124,13 @@ public final T getCurrentBean() {
 		return currentBeanType;
 	}
 
+	public final BeanMetaData<T> getCurrentBeanMetaData() {
+		if ( currentBeanType != null && currentBeanMetaData == null ) {
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValueContext.java]: Why is it done lazily? Seems we'll need it in any case?

File: engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java [hunk 3]
```diff
@@ -118,18 +121,21 @@ public boolean equals(Object o) {
...
-			return false;
-		}
 
 		return true;
 	}
 
+	private static int buildHashCode(ConstraintDescriptorImpl<?> constraintDescriptor, ConstraintLocation location, List<ValueExtractorDescriptor> valueExtractorDescriptors) {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ( constraintDescriptor != null ? constraintDescriptor.hashCode() : 0 );
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java]: Are the null checks really needed?

File: engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractorDescriptor.java [hunk 4]
```diff
@@ -98,7 +98,7 @@ public Key getKey() {
 		return key;
 	}
 
-	public Type getExtractedType() {
+	public Class<?> getExtractedType() {
```
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractorDescriptor.java]: Seems it was wise to limit it to the wildcard parameterization in the spec :)

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java [hunk 5]
```diff
@@ -167,17 +167,19 @@ public ValidatorImpl(ConstraintValidatorFactory constraintValidatorFactory,
 	public final <T> Set<ConstraintViolation<T>> validate(T object, Class<?>... groups) {
 		Contracts.assertNotNull( object, MESSAGES.validatedObjectMustNotBeNull() );
 
-		if ( !beanMetaDataManager.isConstrained( object.getClass() ) ) {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java]: The issue is not `isConstrained` by itself, that's why I kept it and there's still a usage of it.

The issue was that we used `beanMetaDataManager.isConstrained` which got the bean metadata from the concurrent map and then we got again the bean metadata from the concurrent map a few lines below (either really a few lines below or hidden in a method call a few lines below). We should get the bean metadata from the concurrent map only once.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java [hunk 6]
```diff
@@ -68,6 +76,12 @@
 	private final Class<T> rootBeanClass;
 
 	/**
+	 * The metadata of the root bean.
+	 */
+	@Immutable
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: Yeah right, I used it as a marker to say that this value was not supposed to change during the object lifecycle and it was supposed to be "final" as the other ones. I'll remove them.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java [hunk 7]
```diff
@@ -196,6 +214,13 @@ public T getRootBean() {
 		return rootBeanClass;
 	}
 
+	public BeanMetaData<T> getRootBeanMetaData() {
+		if ( rootBeanClass != null && rootBeanMetaData == null ) {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: So in this case, it's an error. It's just that I started with `ValueContext` and then copied the behavior from `ValueContext` to `ValidationContext`. I'll change it here.

File: engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java (hunk 3 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java]: Not really. I'll remove them.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java (hunk 6 above)
Comment (gunnarmorling) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidationContext.java]: Got it. In OGM we have `@EffectivelyImmutable` for a similar purpose (though it applies to initialization of the field during construction time, e.g. a post-construct callback, not lazy-init as done here). If needed we could add it to HV, too.

We don't need to handle those stereotypes with religious believe, it's just kind of an (open) experiment whether its helpful to us or not.

File: engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java (hunk 5 above)
Comment (gsmet) [inline engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java]: Mmmh, but in fact, we can get rid of these too. I got too conservative and wanted to keep the `ValidationContext` initialization after the `hasConstraints` check but it's a bad idea.

I changed it and got rid of `isConstrained()`.

**Comment by gsmet** [comment]
Force pushed an update that addresses all your concerns.

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks, @gsmet!

**Pull request closed without merging**